src/ring_theory/non_zero_divisors.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Devon Tuma
+-/
+
+import group_theory.submonoid.basic
+
+section non_zero_divisors
+
+/-- The submonoid of non-zero-divisors of a `monoid_with_zero` `R`. -/
+def non_zero_divisors (R : Type*) [monoid_with_zero R] : submonoid R :=
+{ carrier := {x | ∀ z, z * x = 0 → z = 0},
+  one_mem' := λ z hz, by rwa mul_one at hz,
+  mul_mem' := λ x₁ x₂ hx₁ hx₂ z hz,
+    have z * x₁ * x₂ = 0, by rwa mul_assoc,
+    hx₁ z $ hx₂ (z * x₁) this }
+
+variables {R A : Type*} [comm_ring R] [integral_domain A]
+
+lemma mul_mem_non_zero_divisors {a b : R} :
+  a * b ∈ non_zero_divisors R ↔ a ∈ non_zero_divisors R ∧ b ∈ non_zero_divisors R :=
```

Comment:
Where do you use that `R` is a `comm_ring`?

Reply:
The forward implication needs `mul_comm a b` at one point, because of the way `non_zero_divisors` is defined. It's really more of a `non_(right)_zero_divisors` if we don't assume the underlying monoid is commutative. Whenever this was originally written it assumed the underlying structures were all commutative, so the left/right distinction didn't matter. To avoid the commutativity here, I think we'd need to strengthen the definition of `non_zero_divisors` to `∀ z, z * x = 0 ∨ x * z = 0 → z = 0`.

Reply:
@dtumad Thanks for clearing that up. Let's not worry about that now.